### PR TITLE
AARCH64 / ARM64 CAPI image build for Ubuntu 22.04

### PR DESCRIPTION
**Pull request by ader1990**
What this PR does / why we need it:

Added example on how to build an ARM64 CAPI image using QEMU.
How to run:

```bash

# under root

apt-get install qemu-system-arm qemu-efi qemu-kvm libvirt-daemon-system -y

# create disk images
pushd /var/lib/libvirt/images/
dd if=/dev/zero of=capi.fd bs=1M count=64
dd if=/dev/zero of=capi-nvmram.fd bs=1M count=64
dd if=/usr/share/qemu-efi-aarch64/QEMU_EFI.fd of=capi.fd conv=notrunc
popd

PACKER_VAR_FILES=arm64.json ARCH=arm64 PACKER_LOG=1 make build-qemu-ubuntu-2204-arm64

# vnc listener is set conveniently to 0.0.0.0:5903 for quicker debug, can be removed later on

# Image should be present at
# ./output/ubuntu-2204-arm64-kube-v1.24.11/ubuntu-2204-arm64-kube-v1.24.11
```

Several parts need to be refactored, any help would be great.

To do's:

1. linux-cloud-tools-virtual debian package does not exist on ARM64 (an exception needs to be added)
2. Replace qemuargs with Packer keywords. I did check on Packer support for ARM64 + QEMU-KVM, but could not see any way to replace the specific qemuargs with Packer equivalent.
3. Cleanup the code after the above have been implemented

Ubuntu 22.04 ARM64 CAPI image with K8s version 1.24 was created and was tested using Tinkerbell Cluster API Provider.

### CODE REVIEW

**Comment by mihaelabalutoiu** [comment]
Also, we need support building ARM64 CAPI images for Flatcar. 
Thank you.

**Comment by jsturtevant** [comment]
> Perhaps we can add it as a separate PR.

makes sense to me, keeps the change set smaller. Would that rely on some of the ansible/goss changes here?

File: images/capi/Makefile [hunk 1]
```diff
@@ -830,11 +831,11 @@ validate-node-ova-local-base-ubuntu-1804: ## Validates Ubuntu 18.04 Base Node OV
 validate-node-ova-local-base-ubuntu-2004: ## Validates Ubuntu 20.04 Base Node OVA w local hypervisor
 validate-node-ova-local-base-ubuntu-2204: ## Validates Ubuntu 22.04 Base Node OVA w local hypervisor
 
-validate-qemu-flatcar: ## Validates Flatcar QEMU image packer config
```
Comment (invidian) [inline images/capi/Makefile:833]: This seem like an unrelated change?

File: images/capi/ansible/roles/setup/tasks/debian.yml [hunk 2]
```diff
@@ -19,7 +19,16 @@
...
-  when: packer_builder_type != "oracle-oci"
+  when: packer_builder_type != "oracle-oci" and ansible_architecture != "aarch64"
+
+- name: Put templated sources.list in place for ARM64
+  template:
+    src: etc/apt/arm64.sources.list.j2
+    dest: /etc/apt/sources.list
+    mode: 0644
+  # OCI Base images have the required apt sources list embedded inside the image, adding the sources list
+  # from this repo leads to build failures(especially in Arm), hence ignoring the step.
```
Comment (invidian) [inline images/capi/ansible/roles/setup/tasks/debian.yml:30]: `Arm` -> `ARM`?

File: images/capi/ansible/roles/setup/tasks/debian.yml [hunk 3]
```diff
@@ -19,7 +19,16 @@
...
+  when: packer_builder_type != "oracle-oci" and ansible_architecture == "aarch64"
```
Comment (invidian) [inline images/capi/ansible/roles/setup/tasks/debian.yml:31]: Given that this is mainly for arm64, would it make sense to swap the conditionals, so we kind of do select + filter? It seem to me it makes it easier to reason about, as it reads as "For ARM64 architecture except builders of type oracle-oci".

File: images/capi/packer/goss/goss-vars.yaml [hunk 4]
```diff
@@ -551,4 +550,5 @@ windows:
         exists: true
         filetype: file
         contains:
-        - "metadata_services=cloudbaseinit.metadata.services.base.EmptyMetadataService"
\ No newline at end of file
+        - "metadata_services=cloudbaseinit.metadata.services.base.EmptyMetadataService"
+
```
Comment (invidian) [inline images/capi/packer/goss/goss-vars.yaml:554]: I think trailing newline is not needed here?

File: images/capi/ansible/roles/node/defaults/main.yml [hunk 5]
```diff
@@ -91,7 +91,6 @@ common_virt_rpms:
 - open-vm-tools
 
 common_virt_debs:
-- linux-cloud-tools-virtual
```
Comment (invidian) [inline images/capi/ansible/roles/node/defaults/main.yml:94]: I'd expect not having this package should cause some validation to fail. If not, perhaps we should have some validation for functionality provided by this package?

Also, if this package is missing for ARM architecture, perhaps we should create an upstream issue for making it available?

File: images/capi/packer/qemu/packer.json [hunk 6]
```diff
@@ -23,12 +23,74 @@
       "net_device": "virtio-net",
       "output_directory": "{{user `output_directory`}}",
       "qemu_binary": "{{user `qemu_binary`}}",
+      "qemuargs": [
```
Comment (invidian) [inline images/capi/packer/qemu/packer.json:26]: Why are all those additional options now needed? If those changes are generic for supporting ARM, could we put them into separate commit with some explanation perhaps, so it is easier to reason about them in the future?

File: images/capi/packer/qemu/qemu-ubuntu-2204-arm64.json [hunk 7]
```diff
@@ -0,0 +1,15 @@
+{
+  "boot_command_prefix": "c<wait>linux /casper/vmlinuz console=ttyAMA0 autoinstall ds='nocloud-net;s=http://{{ .HTTPIP }}:{{ .HTTPPort }}/' --- <enter><wait>initrd /casper/initrd<enter><wait>boot<enter>",
+  "boot_type": "uefi",
+  "build_name": "ubuntu-2204-arm64",
+  "distro_name": "ubuntu",
+  "guest_os_type": "ubuntu-64",
+  "http_directory": "./packer/qemu/files/22.04-arm64",
```
Comment (invidian) [inline images/capi/packer/qemu/qemu-ubuntu-2204-arm64.json:7]: Perhaps this path should point to `packer/qemu/linux/ubuntu/http/22.04.arm64/`, similar to where other http directories are placed?

File: images/capi/packer/qemu/packer.json [hunk 8]
```diff
@@ -23,12 +23,74 @@
...
+        ]
+      ],
       "shutdown_command": "echo '{{user `ssh_password`}}' | sudo -S -E sh -c 'usermod -L {{user `ssh_username`}} && {{user `shutdown_command`}}'",
       "ssh_password": "{{user `ssh_password`}}",
       "ssh_timeout": "2h",
       "ssh_username": "{{user `ssh_username`}}",
       "type": "qemu",
-      "vm_name": "{{user `vm_name`}}"
+      "vm_name": "{{user `vm_name`}}",
+      "vnc_bind_address": "0.0.0.0",
```
Comment (invidian) [inline images/capi/packer/qemu/packer.json:91]: This also seem unrelated to ARM changes, could this be placed in separate commit/PR?

**Review by invidian — COMMENTED** [review]
Awesome work @ader1990, great to see some initial ARM support for CAPI images. Overall PR looks good to me, I've left few comments/questions, which I think clarifying would help.

File: images/capi/Makefile (hunk 1 above)
Comment (ader1990) [reply images/capi/Makefile:833]: yes, will be reverted.

File: images/capi/ansible/roles/node/defaults/main.yml (hunk 5 above)
Comment (ader1990) [reply images/capi/ansible/roles/node/defaults/main.yml:94]: These packages are a bundle up of old days paravirtualized software, like tools + drivers (I think now all drivers are upstreamed already) + services. I think we can create an upstream bug to see what are the alternatives or if these are even required for arm.

File: images/capi/ansible/roles/setup/tasks/debian.yml (hunk 3 above)
Comment (ader1990) [reply images/capi/ansible/roles/setup/tasks/debian.yml:31]: I had to keep backwards compatibility, maybe the conditionals can be aggregated somewhere else?

File: images/capi/packer/goss/goss-vars.yaml (hunk 4 above)
Comment (ader1990) [reply images/capi/packer/goss/goss-vars.yaml:554]: This file did not have consistent file endings with the other files.

File: images/capi/packer/qemu/packer.json (hunk 8 above)
Comment (ader1990) [reply images/capi/packer/qemu/packer.json:91]: These are just for the working example to be usable in a more dev friendly way.

File: images/capi/packer/qemu/packer.json (hunk 6 above)
Comment (ader1990) [reply images/capi/packer/qemu/packer.json:26]: These are the main changes required to be able to start the qemu vm, which are not supported yet as far as I know by packer. Once these changes can be upstreamed in Packer, we can reuse. Without this part, the VM does not even start.

File: images/capi/packer/qemu/qemu-ubuntu-2204-arm64.json (hunk 7 above)
Comment (ader1990) [reply images/capi/packer/qemu/qemu-ubuntu-2204-arm64.json:7]: images/capi/packer/qemu/files/22.04-arm64/user-data should be reusable in some way for any Ubuntu UEFI, I think this can be another PR too.

**Review by ader1990 — COMMENTED** [review]
Mainly, I am thinking to split this PR into at least 3 parts:

1. tooling support for arm64
2. qemu + packer support for arm64
3. Ubuntu related UEFI parts that can be reused by x64 too
